src/main/kotlin/com/safehill/kclient/tasks/outbound/UploadOperationImpl.kt

```diff
@@ -1,64 +1,168 @@
 package com.safehill.kclient.tasks.outbound
 
+import com.safehill.kclient.controllers.LocalAssetsStoreController
+import com.safehill.kclient.models.assets.AssetGlobalIdentifier
+import com.safehill.kclient.models.assets.AssetQuality
+import com.safehill.kclient.models.assets.EncryptedAsset
+import com.safehill.kclient.models.assets.GroupId
+import com.safehill.kclient.models.assets.LocalAsset
+import com.safehill.kclient.models.dtos.AssetOutputDTO
 import com.safehill.kclient.models.users.LocalUser
+import com.safehill.kclient.models.users.ServerUser
 import com.safehill.kclient.network.ServerProxy
-import com.safehill.kclient.tasks.BackgroundTask
+import com.safehill.kcrypto.models.SymmetricKey
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
 import java.util.UUID
 
-public class UploadOperationImpl(
+class UploadOperationImpl(
     val serverProxy: ServerProxy,
     override val listeners: List<UploadOperationListener>,
-) : UploadOperation, BackgroundTask {
+    val encrypter: AssetEncrypterInterface,
+) : UploadOperation {
 
     // TODO: Persist these on disk
-    val outboundQueueItems: MutableList<OutboundQueueItem> = mutableListOf()
+    private val outboundQueueItems: Channel<OutboundQueueItem> = Channel(Channel.UNLIMITED)
+    private val job = Job()
+    private val scope = CoroutineScope(Dispatchers.Default + job)
 
     override val user: LocalUser
         get() = serverProxy.requestor
 
-    override suspend fun upload(outboundQueueItem: OutboundQueueItem) {
+    override suspend fun enqueueUpload(
+        localAsset: LocalAsset,
+        assetQualities: Array<AssetQuality>,
+        groupId: GroupId,
+        recipients: List<ServerUser>
+    ) {
+        scope.launch {
+            val globalIdentifier = UUID.randomUUID().toString()
+            assetQualities.forEach {
+                outboundQueueItems.send(
+                    OutboundQueueItem(
+                        OutboundQueueItem.OperationType.Upload,
+                        it,
+                        localAsset,
+                        globalIdentifier,
+                        groupId,
+                        recipients
+                    )
+                )
+            }
+        }
+    }
+
+    override fun enqueueShare(
+        localAsset: LocalAsset,
+        assetQualities: Array<AssetQuality>,
+        globalIdentifier: AssetGlobalIdentifier,
+        groupId: GroupId,
+        recipients: List<ServerUser>
+    ) {
+        TODO("Not yet implemented")
+    }
+
+    suspend fun upload(outboundQueueItem: OutboundQueueItem) {
+        try {
+            val globalIdentifier = outboundQueueItem.globalIdentifier ?: return
+
+            // *******encrypting*******
+            notifyListenersStartedEncrypting(outboundQueueItem)
```

Comment:
I think we should define an interface for a step that takes a queue item and does something with it. This way this list of steps will be much cleaner.
If input/output are not consistent across all steps we can use generics